Initial work

neilgierman opened the pull request:
Happy path agile implementation. Expect some refactoring to make code more testable. Just a glimpse to see the direction I am taking. More commits to this branch coming in the next few days as I get available spare cycles.

neilgierman committed:
Initial happy path commit. Unit tests coming.

neilgierman committed:
Add some comments and an edge case

neilgierman committed:
Refactor app to make it more testable.

neilgierman committed:
Add unit tests

anton-efimenko commented:
Hi Neil! Thank you for taking your time to work on this. :slightly_smiling_face: We'll get back to you with the comments.

roaet commented:
Howdy! Thank you for working on this. I will make some comments as I go through this throughout the day.

roaet commented:
Given that go standard library container/list is not threadsafe, what can be done to resolve the critical sections?

roaet commented:
If you wanted to scale this, what strategies would you recommend?

neilgierman commented:
> Given that go standard library container/list is not threadsafe, what can be done to resolve the critical sections?

I can add a mutex to the routines that allow me to lock the queue while I am pulling the next entry off, so that if we multithread the queue processor (see next answer), there is not a risk of the next available thread pulling the same list item that the previous thread has started processing but not pulled off the queue.

> If you wanted to scale this, what strategies would you recommend?

Doing the above would allow me to run processPut in a separate thread (go routine) instead of a single thread sequentially running through the queue one by one. After that we could horizontally scale by multiple instances of this service by using containers and container orchestration (k8s/eks ecs).

mkbond commented on cmd/domain-checker/main.go:
```diff
@@ -0,0 +1,19 @@
+)
+
+// Main entrypoint. Everything is in other go files in the same package
+func main() {
+	ctx, cancel := context.WithTimeout(context.Background(), 10*time.Second)
+	defer cancel()
+	a := App{}
+	a.Initialize("administrator","5VhVTxQb272kMsm","cluster0.k6tho.mongodb.net","domain-checker")
+	go a.queueProcessor()
+	a.handleRequests()
+	a.Run()
+	defer a.ReadClient.Disconnect(ctx)
+	defer a.WriteClient.Disconnect(ctx)
+}
```
What would happen if the application were to shut down while messages were still in queue?

neilgierman replied:
Without a persistent store backing the queue, if the process dies, the queue is lost.

neilgierman replied:
Although not asked, a resolution to that is something like RabbitMQ. It does make the system a little complex but not only allows a process to die without losing the queue, but also allow multiple instances to be brought up and read from the same queue. RabbitMQ is about the lightest weight solution that I have consumed in other projects.

anton-efimenko replied:
@neilgierman how could you defend @mkbond's argument without adding complexity (such as RabbitMQ) into the system?

neilgierman replied:
In order to save the queue across process restarts, we would need to persist the list somewhere outside of process memory. Without diving into RabbitMQ, we could persist the list to a file (making sure to lock and unlock it to prevent multiple threads from trying to write to the file at the same time). If the process dies and is restarted on the same machine, the replacement process can load the previous queue from the file and resume.

anton-efimenko replied:
What about the use-case when a process doesn't disappear, but simply receives a SIGTERM?

neilgierman replied:
If you get a SIGTERM, you probably want to stop the handler from getting new requests and process the queue that remains so that those queued entries will still be processed. Then once the queue is empty you can kill the process.

anton-efimenko replied:
Yup! :+1:

roaet commented:
If `processPut` is run by multiple threads or by multiple instances of the service what are some issues that you may face?

mkbond commented on cmd/domain-checker/app.go:
```diff
@@ -0,0 +1,76 @@
+package main
```
Any downside to having everything in the same package?

neilgierman replied:
With everything being in the same package, it is not reusable by other code. We could put more of the back-end code into a different package so that if another project wanted to reuse this code they could "go get" the backend library and write their own front-end. It also makes it a little difficult with code maintenance everything being in the same directory as the code gets larger.

mkbond commented on cmd/domain-checker/app.go:
```diff
@@ -0,0 +1,76 @@
+	"go.mongodb.org/mongo-driver/mongo/readpref"
+)
+
+type App struct {
+	Router *mux.Router
+	ReadClient *mongo.Client
+	WriteClient *mongo.Client
+	ReadDb *mongo.Database
+	WriteDb *mongo.Database
+	Queue *list.List
+}
```
Does everything here need to be public?

mkbond replied:
Additionally is there a better way to format these?

neilgierman replied:
> Does everything here need to be public?

No. With everything in the same package, they don't need to be public.

> Additionally is there a better way to format these?

I could make the structure more than one level such as:
```
type App struct {
    Rest RestHelper
    DB DatabaseHelper
    Queue QueueHelper
}

type RestHelper struct {
    Router *mux.Router
}

type DatabaseHelper struct {
    ReadClient *mongo.Client
    WriteClient *mongo.Client
    ReadDb *mongo.Database
    WriteDb *mongo.Database
}

type QueueHelper struct {
    Queue *list.List
    Lock *sync.Mutex
}
```

mkbond commented on cmd/domain-checker/main.go:
```diff
@@ -0,0 +1,19 @@
+	ctx, cancel := context.WithTimeout(context.Background(), 10*time.Second)
+	defer cancel()
+	a := App{}
+	a.Initialize("administrator","5VhVTxQb272kMsm","cluster0.k6tho.mongodb.net","domain-checker")
```
How would we avoid putting credentials directly in the code?

neilgierman replied:
We can use many different methods; environment variables (either at runtime, or at compile time using GitHub Secrets), and configuration files. There are probably more, but those are what I have used in the past.

mkbond reviewed (commented):
Hey Neil! Thanks to for taking our coding challenge. Left a few questions 🙂

neilgierman commented:
> If `processPut` is run by multiple threads or by multiple instances of the service what are some issues that you may face?

First we would need to make sure the queue is protected so multiple threads won't be processing the same queue entry. Also inside processPut, we do a getDomain to see if we need to make an InsertOne or UpdateOne call. We could have a race condition where 2 threads do a getDomain and return not found and setup to do an insert for both. One way to address that is rearchitect the queue slightly so that if there is already an entry in the queue for a specific domain, add another to the queue item count instead of adding a duplicate queue entry, something like:
```
type queueEntry struct {
    action string
    domain string
    count int
}
```
Then in the respective service::queue[Delivered|Bounced] have a bit more logic to search the queue for an existing record and if found, increase that record's count member instead of adding a new record.

roaet commented:
If you were to rearchitect the queue, what structure would you use? Also, are there any other race conditions you can see with multiple DB writers?

neilgierman commented:
> If you were to rearchitect the queue, what structure would you use? Also, are there any other race conditions you can see with multiple DB writers?

If we switch to RabbitMQ, then I won't be able to look at the entire queue and consolidate duplicate action/domain pairs so the queueEntry would stay the same. I would just change the queue implementation from a containers/list to RabbitMQ-hosted. This makes unit tests a little more complicated however still do-able with projects such as https://github.com/NeowayLabs/wabbit.

As far as other issues with multiple DB writers, there is a concern about trying to update the counts for a specific domain name by 2 different routines. We could in theory lose the value of the first writer when the second writer comes in. In other languages, I have used named mutexes to only allow 1 thread to operate on a named resource (domain name in this case) at a time. From initial research, named mutexes are not a part of the base go library (sync), however other projects have implemented the concept such as https://github.com/juju/mutex

neilgierman committed:
Switch from internal container/list to RabbitMQ

neilgierman committed:
Switch to a local MongoDB instance and config file

neilgierman committed:
Refactor processPut

First use FindOneAndUpdate with Upsert=true so we don't have a
different branch for insert vs update.
Also use the built-in increment function in MongoDB so we
don't have to read to get the current counts and a separate write to
store the new counts.

neilgierman committed:
Reorganize App config so not flat.

neilgierman committed:
Create docker system and stress test

Create Dockerfile for the app itself
Create docker-compose Yaml for the entire system
Create stress test PowerShell script to confirm system operates
as expected with 2 apps targeting the same queue and same DB

Tested on Windows host.

neilgierman committed:
Make Alpine test service

Just for testing.
Added alpine test client.
Configured to hang on service start
Once started, enter it by `docker exec -it domain-checker_client_1 sh`
Once inside,  create stress-test.sh with contents and execute.

neilgierman committed:
Enhance docker/test infrastructure

neilgierman committed:
First attempt GitHub Actions

neilgierman committed:
Update permissions on workflow

neilgierman committed:
Add Build Status badge

neilgierman committed:
Add Code Coverage to Actions

anton-efimenko commented on cmd/domain-checker/database.go:
```diff
@@ -0,0 +1,87 @@
+	case "delivered":
+		update = bson.M{
+			"$inc": bson.M{
+				"deliveredCount":1,
```
Great job iterating on the solution!

Let's imagine we want to process billions of these per second which will put quite a bit of strain on the db. Let's assume we can't replace a database or add sharding. What can we do on the application side to improve the performance without changing the infrastructure?

P.S.
Let me know if you need a hint.

anton-efimenko replied:
cc @neilgierman

anton-efimenko replied:
cc @neilgierman

neilgierman replied:
Depending on the SLA, we could batch these writes. As long as we are allowed to have a certain period of time between a PUT of a domain and a GET of the same domain getting updated counts, we could hold on to a certain number of write requests in a different queue and when a certain queue size, or time has elapsed we can push those writes all at once. That will reduce the connection stress on the DB connections somewhat but trade off not having up-to-the-nanosecond records for reads.

Open to other hints.

anton-efimenko replied:
Other than the "trade off not having up-to-the-nanosecond records for reads" what would be other downsides? Can we afford it?

neilgierman replied:
You have one more mutex to keep track of so that you aren't trying to process the write queue while you are adding to it. You have more memory consumed that can get rather large depending on to many write transactions you queue up before you process it. If you don't back up the queue with some sort of storage, you have a risk of losing the write queue if the process ends.

anton-efimenko replied:
The last point was the one I specifically tried to walk you to :slightly_smiling_face: 
That being said I would argue it's a fair trade-off because we don't have to be precise with the calculations, from the product perspective it's a heuristic-based algorithm to begin with.

neilgierman committed:
Batch database writes

roaet commented on cmd/domain-checker/database.go:
```diff
@@ -0,0 +1,116 @@
+	batchUpdates = append(batchUpdates, batchUpdate)
+	log.Print(batchUpdate, " added to batchUpdates")
+	log.Print("batchUpdates is ", len(batchUpdates))
+	a.dbConfig.writeDbLock.Unlock()
```
The amount of instructions between mutex lock and unlock is fairly high. What could be another way to do this with a smaller critical section?

neilgierman replied:
I can move the lock and unlock just around the append instruction since that is really the only instruction that needs to be protected. 

Also, since go channels block by default, I should be able to change the mutex to a channel and update the relevant area to deal with signals to/from the channel instead of a lock/unlock. However, channels have some internal overhead that would have to be weighed for a simple lock on a single resource.

mkbond commented on cmd/domain-checker/database.go:
```diff
@@ -0,0 +1,116 @@
+	bulkOption.SetOrdered(true)
+
+	for {
+		if time.Since(a.dbConfig.lastUpdated) > (2 * time.Second) && len(batchUpdates) > 0 {
```
What would you think about using a `time.Ticker` or `time.After()` here?

neilgierman replied:
Yes. Moving to time.After to be more "go" and also save a small bit of memory in App structure.

neilgierman committed:
Lock cleanups

Move lock/unlock in processPut closer together
Move from lastUpdated to time.After for batch writer monitor

neilgierman committed:
Change GET output from raw to json.

neilgierman committed:
Add a few more unit tests.

neilgierman committed:
Make a load balancing web front end
